### PR TITLE
Generate: Update docs regarding reusing `past_key_values` in `generate`

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Due to a [recent PR](https://github.com/huggingface/transformers/pull/25086), which enables `generate` to return `past_key_values`, we can now reuse `past_key_values` to speed up multi-round conversations with LLMs. This was already described in the [LLM optimization docs](https://huggingface.co/docs/transformers/llm_tutorial_optimization), even though there was no corresponding code back when it was written.

This PR updates the doc with a code example of how to return and reuse `past_key_values`, polishing a few nits encountered along the way.

Related issue: https://github.com/huggingface/transformers/issues/27546

### CODE REVIEW

File: docs/source/en/llm_tutorial_optimization.md [hunk 1]
```diff
@@ -22,7 +22,7 @@ The crux of these challenges lies in augmenting the computational and memory cap
 
 In this guide, we will go over the effective techniques for efficient LLM deployment:
 
-1.  **Lower Precision**: Research has shown that operating at reduced numerical precision, namely [8-bit and 4-bit](./main_classes/quantization.md) can achieve computational advantages without a considerable decline in model performance.
+1.  **Lower Precision:** Research has shown that operating at reduced numerical precision, namely [8-bit and 4-bit](./main_classes/quantization.md) can achieve computational advantages without a considerable decline in model performance.
```
Comment (gante) [inline docs/source/en/llm_tutorial_optimization.md:25]: (odd one out)

File: docs/source/en/llm_tutorial_optimization.md [hunk 2]
```diff
@@ -58,7 +58,7 @@ As of writing this document, the largest GPU chip on the market is the A100 & H1
...
 
 Naive pipeline parallelism is supported out of the box. For this, simply load the model with `device="auto"` which will automatically place the different layers on the available GPUs as explained [here](https://huggingface.co/docs/accelerate/v0.22.0/en/concept_guides/big_model_inference).
-Note, however that while very effective, this naive pipeline parallelism does not tackle the issues of GPU idling. For this more advanced pipeline parallelism is required as explained [here](https://huggingface.co/docs/transformers/v4.34.0/en/perf_train_gpu_many#naive-model-parallelism-vertical-and-pipeline-parallelism).
+Note, however that while very effective, this naive pipeline parallelism does not tackle the issues of GPU idling. For this more advanced pipeline parallelism is required as explained [here](https://huggingface.co/docs/transformers/en/perf_train_gpu_many#naive-model-parallelism-vertical-and-pipeline-parallelism).
```
Comment (gante) [inline docs/source/en/llm_tutorial_optimization.md:61]: (link now points to the latest stable version)

File: docs/source/en/llm_tutorial_optimization.md [hunk 3]
```diff
@@ -286,7 +286,7 @@ If GPU memory is not a constraint for your use case, there is often no need to l
 For more in-detail usage information, we strongly recommend taking a look at the [Transformers Quantization Docs](https://huggingface.co/docs/transformers/main_classes/quantization#general-usage).
 Next, let's look into how we can improve computational and memory efficiency by using better algorithms and an improved model architecture.
 
-# 2. Flash Attention
+## 2. Flash Attention
```
Comment (gante) [inline docs/source/en/llm_tutorial_optimization.md:289]: The missing `#` was resulting in a poorly rendered TOC on the right-hand side (see [here](https://huggingface.co/docs/transformers/llm_tutorial_optimization))

File: docs/source/en/llm_tutorial_optimization.md [hunk 4]
```diff
@@ -662,7 +664,15 @@ Using the key-value cache has two advantages:
 
 > One should *always* make use of the key-value cache as it leads to identical results and a significant speed-up for longer input sequences. Transformers has the key-value cache enabled by default when making use of the text pipeline or the [`generate` method](https://huggingface.co/docs/transformers/main_classes/text_generation).
 
-Note that the key-value cache is especially useful for applications such as chat where multiple passes of auto-regressive decoding are required. Let's look at an example.
+Tip warning={true}>
```
Comment (gante) [inline docs/source/en/llm_tutorial_optimization.md:667]: Added this warning with the recent finding that using caches produces tiny fluctuations in the model, so users can find this info more easily :)

File: docs/source/en/llm_tutorial_optimization.md [hunk 5]
```diff
@@ -662,7 +664,15 @@ Using the key-value cache has two advantages:
...
+
+Note that, despite our advice to use key-value caches, your LLM output may be slightly different when you use them. This is a property of the matrix multiplication kernels themselves -- you can read more about it [here](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535).
+
+</Tip>
+
+#### 3.2.1 Multi-round conversation
```
Comment (gante) [inline docs/source/en/llm_tutorial_optimization.md:673]: No text was modified in this part, only added: from "In `transformers`, a `generate` call" to "for the attention layer!"

**Comment by gante** [comment]
The right-handed TOC is now fully functional!

Screenshot 2023-11-20 at 18 31 56

File: docs/source/en/llm_tutorial_optimization.md [hunk 6]
```diff
@@ -662,7 +664,15 @@ Using the key-value cache has two advantages:
 
 > One should *always* make use of the key-value cache as it leads to identical results and a significant speed-up for longer input sequences. Transformers has the key-value cache enabled by default when making use of the text pipeline or the [`generate` method](https://huggingface.co/docs/transformers/main_classes/text_generation).
 
-Note that the key-value cache is especially useful for applications such as chat where multiple passes of auto-regressive decoding are required. Let's look at an example.
+<Tip warning={true}>
+
+Note that, despite our advice to use key-value caches, your LLM output may be slightly different when you use them. This is a property of the matrix multiplication kernels themselves -- you can read more about it [here](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535).
+
+</Tip>
```
Comment (ArthurZucker) [inline docs/source/en/llm_tutorial_optimization.md:671]: 🚀

**Review by ArthurZucker — APPROVED** [review]
Great including the padding left / right comment! 🔥

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot!

**Pull request merged**